extras/rya.indexing.pcj/src/main/java/org/apache/rya/indexing/pcj/storage/mongo/MongoPcjStorage.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.rya.api.model.VisibilityBindingSet;
+import org.apache.rya.api.utils.CloseableIterator;
+import org.apache.rya.indexing.pcj.storage.PCJIdFactory;
+import org.apache.rya.indexing.pcj.storage.PcjMetadata;
+import org.apache.rya.indexing.pcj.storage.PrecomputedJoinStorage;
+import org.apache.rya.mongodb.instance.MongoRyaInstanceDetailsRepository;
+import org.openrdf.query.BindingSet;
+
+import com.mongodb.MongoClient;
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A mongo backed implementation of {@link PrecomputedJoinStorage}.
+ */
+@DefaultAnnotation(NonNull.class)
+public class MongoPcjStorage implements PrecomputedJoinStorage {
+    public static final String PCJ_COLLECTION_NAME = "pcjs";
+    // Used to update the instance's metadata.
+    private final MongoRyaInstanceDetailsRepository ryaDetailsRepo;
+
+    private final String ryaInstanceName;
+
+    // Factories that are used to create new PCJs.
+    private final PCJIdFactory pcjIdFactory = new PCJIdFactory();
+
+    private final MongoPcjDocuments pcjDocs;
+
+    /**
+     * Constructs an instance of {@link MongoPcjStorage}.
+     *
+     * @param client - The {@link MongoClient} that will be used to connect to Mongodb. (not null)
+     * @param ryaInstanceName - The name of the RYA instance that will be accessed. (not null)
+     */
+    public MongoPcjStorage(final MongoClient client, final String ryaInstanceName) {
+        requireNonNull(client);
+        this.ryaInstanceName = requireNonNull(ryaInstanceName);
+        pcjDocs = new MongoPcjDocuments(client, ryaInstanceName);
+        ryaDetailsRepo = new MongoRyaInstanceDetailsRepository(client, ryaInstanceName);
+    }
+
+    @Override
+    public String createPcj(final String sparql) throws PCJStorageException {
+        requireNonNull(sparql);
+
+        // Update the Rya Details for this instance to include the new PCJ
+        // table.
+        final String pcjId = pcjIdFactory.nextId();
+
+        try {
+            new RyaDetailsUpdater(ryaDetailsRepo).update(originalDetails -> {
+                // Create the new PCJ's details.
+                final PCJDetails.Builder newPcjDetails = PCJDetails.builder().setId(pcjId);
+
+                // Add them to the instance's details.
+                final RyaDetails.Builder mutated = RyaDetails.builder(originalDetails);
+                mutated.getPCJIndexDetails().addPCJDetails(newPcjDetails);
+                return mutated.build();
+            });
+        } catch (final RyaDetailsRepositoryException | CouldNotApplyMutationException e) {
+            throw new PCJStorageException(String.format("Could not create a new PCJ for Rya instance '%s' "
+                    + "because of a problem while updating the instance's details.", ryaInstanceName), e);
+        }
+
+        // Create the objectID of the document to house the PCJ results.
+        pcjDocs.createPcj(pcjId, sparql);
```

Comment:
See, right here you're calling it a pcjId which is in line with the rest of the application, but then inside of that class you're calling it a pcjName and changing the meaning of pcjId. That's inconsistent.

Reply:
I'll change it to ID everywehre